pkg/cmd/cli/cmd/export.go

```diff
@@ -0,0 +1,167 @@
+	cmd.Flags().Bool("exact", false, "Preserve fields that may be cluster specific, such as service portalIPs or generated names")
+	cmd.Flags().Bool("raw", false, "If true, do not alter the resources in any way after they are loaded.")
+	cmd.Flags().StringP("selector", "l", "", "Selector (label query) to filter on")
+	cmd.Flags().Bool("all", false, "select all resources in the namespace of the specified resource types")
+	cmd.Flags().VarP(&filenames, "filename", "f", "Filename, directory, or URL to file to use to edit the resource.")
+	cmdutil.AddPrinterFlags(cmd)
+	return cmd
+}
+
+func RunExport(f *clientcmd.Factory, exporter Exporter, in io.Reader, out io.Writer, cmd *cobra.Command, args []string, filenames util.StringList) error {
+	selector := cmdutil.GetFlagString(cmd, "selector")
+	all := cmdutil.GetFlagBool(cmd, "all")
+	exact := cmdutil.GetFlagBool(cmd, "exact")
+	asTemplate := cmdutil.GetFlagBool(cmd, "as-template")
+	raw := cmdutil.GetFlagBool(cmd, "raw")
+	if exact && raw {
+		return cmdutil.UsageError(cmd, "--exact and --raw may not both be specified")
+	}
+
+	clientConfig, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+	outputVersion := cmdutil.OutputVersion(cmd, clientConfig.Version)
+
+	cmdNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	mapper, typer := f.Object()
+	b := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(filenames...).
+		SelectorParam(selector).
+		ResourceTypeOrNameArgs(all, args...).
+		Flatten()
+
+	one := false
+	infos, err := b.Do().IntoSingular(&one).Infos()
+	if err != nil {
+		return err
+	}
+
+	if len(infos) == 0 {
+		return fmt.Errorf("no resources found - nothing to export")
+	}
+
+	if !raw {
+		newInfos := []*resource.Info{}
+		errs := []error{}
+		for _, info := range infos {
+			if err := exporter.Export(info.Object, exact); err != nil {
+				if err == ErrExportOmit {
+					continue
+				}
+				errs = append(errs, err)
+			}
+			newInfos = append(newInfos, info)
+		}
+		if len(errs) > 0 {
+			return utilerrors.NewAggregate(errs)
+		}
+		infos = newInfos
+	}
+
+	var result runtime.Object
+	if asTemplate {
+		objects, err := resource.AsVersionedObjects(infos, outputVersion)
+		if err != nil {
+			return err
+		}
+		template := &templateapi.Template{
+			Objects: objects,
+		}
+		result, err = kapi.Scheme.ConvertToVersion(template, outputVersion)
+		if err != nil {
+			return err
+		}
+	} else {
+		object, err := resource.AsVersionedObject(infos, !one, outputVersion)
+		if err != nil {
+			return err
+		}
+		result = object
+	}
+
+	// use YAML as the default format
+	outputFormat := cmdutil.GetFlagString(cmd, "output")
+	templateFile := cmdutil.GetFlagString(cmd, "template")
+	if len(outputFormat) == 0 && len(templateFile) != 0 {
+		outputFormat = "template"
+	}
+	if len(outputFormat) == 0 {
+		outputFormat = "yaml"
```

Comment:
Over time I expect a lot of our default output to become YAML.  It's YAML here specifically because this is something we want end users to edit and manage.

----- Original Message -----
> > +	} else {
> > +		object, err := resource.AsVersionedObject(infos, !one, outputVersion)
> > +		if err != nil {
> > +			return err
> > +		}
> > +		result = object
> > +	}
> > +
> > +	// use YAML as the default format
> > +	outputFormat := cmdutil.GetFlagString(cmd, "output")
> > +	templateFile := cmdutil.GetFlagString(cmd, "template")
> > +	if len(outputFormat) == 0 && len(templateFile) != 0 {
> > +		outputFormat = "template"
> > +	}
> > +	if len(outputFormat) == 0 {
> > +		outputFormat = "yaml"
> 
> I was wondering if we slowly migrate from json to yaml as the default format
> for our commands
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2910/files#r31897736